src/NHapi.Base/Parser/XMLParser.cs

```diff
@@ -25,796 +25,828 @@
+            var all = segmentElement.ChildNodes;
+            for (var i = 0; i < all.Count; i++)
+            {
+                var elementName = all.Item(i).Name;
+                if (Convert.ToInt16(all.Item(i).NodeType) == (short)XmlNodeType.Element && !done.Contains(elementName))
+                {
+                    done.Add(elementName);
+
+                    var index = elementName.IndexOf('.');
+                    if (index >= 0 && elementName.Length > index)
+                    {
+                        // properly formatted element
+                        var fieldNumString = elementName.Substring(index + 1);
+                        var fieldNum = int.Parse(fieldNumString);
+                        ParseReps(segmentObject, segmentElement, elementName, fieldNum);
+                    }
+                    else
+                    {
+                        Log.Debug("Child of segment " + segmentObject.GetStructureName() + " doesn't look like a field: " + elementName);
+                    }
+                }
+            }
+
+            // set data type of OBX-5
+            if (segmentObject.GetType().FullName.IndexOf("OBX") >= 0)
+            {
+                Varies.FixOBX5(segmentObject, Factory);
+            }
+        }
+
+        /// <summary> Populates the given Element with data from the given Segment, by inserting
+        /// Elements corresponding to the Segment's fields, their components, etc.  Returns
+        /// true if there is at least one data value in the segment.
+        /// </summary>
+        public virtual bool Encode(ISegment segmentObject, XmlElement segmentElement)
+        {
+            var hasValue = false;
+            var n = segmentObject.NumFields();
+            for (var i = 1; i <= n; i++)
+            {
+                var name = MakeElementName(segmentObject, i);
+                var reps = segmentObject.GetField(i);
+                for (var j = 0; j < reps.Length; j++)
+                {
+                    var newNode = segmentElement.OwnerDocument.CreateElement(name);
+                    var componentHasValue = Encode(reps[j], newNode);
+                    if (componentHasValue)
+                    {
+                        try
+                        {
+                            segmentElement.AppendChild(newNode);
+                        }
+                        catch (Exception e)
+                        {
+                            throw new HL7Exception("DOMException encoding Segment: ", ErrorCode.APPLICATION_INTERNAL_ERROR, e);
+                        }
+
+                        hasValue = true;
+                    }
+                }
+            }
+
+            return hasValue;
+        }
+
+        /// <summary> Populates the given Type object with data from the given XML Element.</summary>
+        public virtual void Parse(IType datatypeObject, XmlElement datatypeElement)
+        {
+            if (datatypeObject is Varies)

```

Comment:
this would potentially use the new switch expression (if supported by net35)

Reply:
left a `TODO` comment

Reply:
cool, but it's a switch expression not a switch statement 😋